test/topology_custom/test_multidc.py

```diff
@@ -0,0 +1,209 @@
+    cf = "test_cf"
+    table_name = "test_table_name"
+    columns = [Column("name", TextType), Column("value", TextType)]
+    logger.info("Create two servers in different DC's")
+    for i in [1, 1, 2, 2]:
+        s_info = await manager.server_add(
+            config=CONFIG,
+            property_file={"dc": f"dc{i}", "rack": "myrack"},
+        )
+        logger.info(s_info)
+    conn = manager.get_cql()
+    logger.info(conn.execute('desc cluster').one())
+    random_tables = RandomTables(request.node.name, manager, ks, 2)
+    await random_tables.add_table(ncolumns=2, columns=columns, pks=1, name=table_name)
+    conn.execute(f"USE {ks}")
+    conn.execute(
+        f"CREATE COLUMNFAMILY {cf} ( key varchar, c varchar, v varchar, PRIMARY KEY(key, c)) WITH comment='test cf'"
+    )
+
+    logger.info("Perform insert/overwrite")
+    putget(conn, cf, ks)
+
+
+@pytest.mark.asyncio
+async def test_query_dc_with_rf_0_does_not_crash_db(request: pytest.FixtureRequest, manager: ManagerClient):
+    """Test querying dc with CL=LOCAL_QUORUM when RF=0 for this dc, does not crash the node and returns None
+    Covers https://github.com/scylladb/scylla/issues/8354"""
+    servers = []
+    ks = "test_ks"
+    table_name = "test_table_name"
+    expected = ["k1", "value1"]
+    dc_replication = {'dc2': 0}
+    columns = [Column("name", TextType), Column("value", TextType)]
+
+    for i in [1, 2]:
+        servers.append(await manager.server_add(
+            config=CONFIG,
+            property_file={"dc": f"dc{i}", "rack": "myrack"},
+        ))
+
+    dc1_connection = manager.connect_to_node(servers[0])
+    dc2_connection = manager.connect_to_node(servers[1])
+    random_tables = RandomTables(request.node.name, manager, ks, 1, dc_replication)
+    await random_tables.add_table(ncolumns=2, columns=columns, pks=1, name=table_name)
+    dc1_connection.execute(
+        f"INSERT INTO  {ks}.{table_name} ({columns[0].name}, {columns[1].name}) VALUES ('{expected[0]}', '{expected[1]}');")
+    query_cl_one = SimpleStatement(f"SELECT * from {ks}.{table_name};", consistency_level=ConsistencyLevel.ONE)
+    query_cl_local_quorum = SimpleStatement(f"SELECT * from {ks}.{table_name};",
+                                            consistency_level=ConsistencyLevel.LOCAL_QUORUM)
+    nodetool.flush(dc1_connection, "{ks}.{table_name}")
+    first_node_results = list(dc1_connection.execute(query_cl_one).one())
+    second_node_result = dc2_connection.execute(query_cl_local_quorum).one()
+
+    assert first_node_results == expected, \
+        f"Expected {expected} from {query_cl_one.query_string}, but got {first_node_results}"
+    assert second_node_result is None, \
+        f"Expected no results from {query_cl_local_quorum.query_string}, but got {second_node_result}"
+
+
+def _put_with_overwrite(session, nb_keys, cf, ks, cl=ConsistencyLevel.QUORUM):
+    """
+    Put 100 rows with for specified keys. Then overwrite partially existing rows with new values
```

Comment:
It seems the code below actually adds nb_keys * 100 rows?
What is nb_keys about at all?

Personally, I would love for this function to disappear. If I understand correctly, you only use it once in the code. So just insert it where its needed, and add a comment *exactly* what it is writing.

Also, is all this elaborate code to write 100 rows and then rewrite some of them, etc., needed for this test? Does it test anything that just writing 10 rows once wouldn't have tested?

Reply:
Modified and simplified code.
This is a migrated test, and I think the idea behind it is to check that there are no issues with updating the values in a multi-DC environment.